File: android-core/src/androidTest/java/com/mparticle/internal/database/tables/mp/BaseTableTest.java
package com.mparticle.internal.database.tables.mp;

import android.database.sqlite.SQLiteDatabase;
import android.support.test.InstrumentationRegistry;

import com.mparticle.testutils.BaseCleanInstallEachTest;
import com.mparticle.internal.database.BaseDatabase;
import com.mparticle.internal.database.services.SQLiteOpenHelperWrapper;

import org.junit.Before;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class BaseTableTest extends BaseCleanInstallEachTest {
    public static final String DB_NAME = "test_database";
    CountDownLatch timer = new CountDownLatch(1);

    protected void runTest(SQLiteOpenHelperWrapper helper) throws InterruptedException {
        runTest(helper, 6);
    }

    protected void runTest(SQLiteOpenHelperWrapper helper, int oldVersion) throws InterruptedException {
        InstrumentationRegistry.getTargetContext().deleteDatabase(DB_NAME);
        SQLiteDatabase baseDatabase = new BaseDatabase(helper, DB_NAME, timer, oldVersion).getWritableDatabase();
        timer.await(5, TimeUnit.SECONDS);
        baseDatabase = new BaseDatabase(helper, DB_NAME, timer, MParticleDatabaseHelper.DB_VERSION).getWritableDatabase();
        timer.await(5, TimeUnit.SECONDS);
        InstrumentationRegistry.getTargetContext().deleteDatabase(DB_NAME);
    }

    @Before
    public final void before() throws Exception {
        deleteTestingDatabase();
    }

    protected void deleteTestingDatabase() {
        InstrumentationRegistry.getTargetContext().deleteDatabase(DB_NAME);
    }
}

File: android-core/src/androidTest/kotlin/com.mparticle/internal/database/tables/BaseTableTest.kt
package com.mparticle.internal.database.tables

import androidx.test.platform.app.InstrumentationRegistry
import com.mparticle.internal.database.TestSQLiteOpenHelper
import com.mparticle.internal.database.services.SQLiteOpenHelperWrapper
import com.mparticle.testutils.BaseCleanInstallEachTest
import com.mparticle.testutils.MPLatch
import org.junit.Before
import java.util.concurrent.CountDownLatch

open class BaseTableTest : BaseCleanInstallEachTest() {
    var onCreateLatch: CountDownLatch = MPLatch(1)
    var onUpgradeLatch: CountDownLatch = MPLatch(1)

    @Throws(InterruptedException::class)
    protected fun runTest(helper: SQLiteOpenHelperWrapper?, oldVersion: Int = 6) {
        InstrumentationRegistry.getInstrumentation().targetContext.deleteDatabase(DB_NAME)
        var openHelper = helper?.let { TestSQLiteOpenHelper(it, DB_NAME, oldVersion) }
        openHelper?.writableDatabase
        openHelper?.onCreateLatch?.await()
        openHelper =
            helper?.let { TestSQLiteOpenHelper(it, DB_NAME, MParticleDatabaseHelper.DB_VERSION) }
        openHelper?.writableDatabase
        if (oldVersion < MParticleDatabaseHelper.DB_VERSION) {
            openHelper?.onUpgradeLatch?.await()
        }
        InstrumentationRegistry.getInstrumentation().targetContext.deleteDatabase(DB_NAME)
    }

    @Before
    @Throws(Exception::class)
    fun before() {
        deleteTestingDatabase()
    }

    protected fun deleteTestingDatabase() {
        InstrumentationRegistry.getInstrumentation().targetContext.deleteDatabase(DB_NAME)
    }

    companion object {
        const val DB_NAME = "test_database"
    }
}
